# Анализ метаданных источников Hugging Face и Kaggle

Входяшие файлы: извлеченные и обработанные (reaper/processed) метаданные датасетов по схеме данных dateno. 



## 

In [50]:
#установка соединения с duckdb
import duckdb as dd
conn = dd.connect()

## Hugging Face

Дата загрузки изучаемых файлов: 24 апреля 2025 (2025-04-24). 

In [ ]:
#чтение и объединение всех parquet файлов в одну таблицу hf_table (из локальной папки)

conn.sql("""
CREATE TABLE hf_table AS
    SELECT * 
    FROM read_parquet("C:\\Users\\Kseniya\\dateno\\analyze may\\processed\\2025-04-24\\cdi00001985\\hugging_face\\*.parquet", union_by_name=TRUE);
        """)

FloatProgress(value=0.0, layout=Layout(width='auto'), style=ProgressStyle(bar_color='black'))

In [53]:
#просмотр таблицы в колоночном виде

conn.sql("""
SELECT * 
FROM hf_table;
        """)

┌──────────────────────────────────────────────────────────────────┬─────────────┬────────────────────────────────────────────────────────────────────────────────────────────────────────────┬────────────────────────────────────────────────────────────────────────────────────────────────────────────┬──────────────────────────────────────────────────────────────────────────────────┬─────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────

In [54]:
# преобразование таблицы в датафрем и подсчет числа строк

df_hugging_process = conn.sql("""
SELECT * 
FROM hf_table;
        """). df()


len(df_hugging_process)

363175

Всего **363 175** датасетов содержится в таблице по Hugging Face.

In [ ]:
# просмотр первых пяти строчек таблицы
df_hugging_process.head()

id       cdi_id  \
0  dcf198fc8dcf6da94d7f533f5ace8d0e2120a410c750e0...  cdi00001985   
1  11e84d9279f1d25a59c401d477522abeaa223002ae8d4c...  cdi00001985   
2  5ec5aed1c1e83443c9d35978d8c0e0d04302386997223b...  cdi00001985   
3  082511ccef8ec8659cfd0ac89f077f1ce7842d889bca3b...  cdi00001985   
4  c724bd0a5963fc665ac7a74caed418e469bbd5918b20ea...  cdi00001985   

                                              int_id  \
0                                         s-nlp/sqwd   
1            lbourdois/radios_et_podcasts_en_ligne_8   
2                                 tasksource/SpaRTUN   
3  somerandomguyontheweb/multilingual-pl-bert-be-...   
4                                   pvrancx/tyk2_fep   

                                               title  \
0                                         s-nlp/sqwd   
1            lbourdois/radios_et_podcasts_en_ligne_8   
2                                 tasksource/SpaRTUN   
3  somerandomguyontheweb/multilingual-pl-bert-be-...   
4                                   pvrancx/tyk2_fep   

                                          short_text  \
0  SimpleQuestion dataset based on Wikidata with ...   
1  Les données présentes dans ce répertoire ont é...   
2  Dataset Card for "SpaRTUN"https://github.com/H...   
3  This is a replacement for the folder https://h...   
4   Molecular dataset: 10,000 TYK2 inhibitors (SM...   

                                         description  \
0  SimpleQuestion dataset based on Wikidata with ...   
1  Les données présentes dans ce répertoire ont é...   
2  Dataset Card for "SpaRTUN"https://github.com/H...   
3  This is a replacement for the folder https://h...   
4   Molecular dataset: 10,000 TYK2 inhibitors (SM...   

                                                 url        date_created  \
0         https://huggingface.co/datasets/s-nlp/sqwd 2023-12-30 10:19:47   
1  https://huggingface.co/datasets/lbourdois/radi... 2023-12-30 15:38:13   
2  https://huggingface.co/datasets/tasksource/Spa... 2023-12-30 19:08:46   
3  https://huggingface.co/datasets/somerandomguyo... 2023-12-30 19:46:26   
4   https://huggingface.co/datasets/pvrancx/tyk2_fep 2023-12-30 20:35:57   

         date_changed frequency  ...  \
0 2023-12-30 10:25:11   Unknown  ...   
1 2025-04-05 17:42:42   Unknown  ...   
2 2024-01-03 09:02:29   Unknown  ...   
3 2023-12-31 09:00:52   Unknown  ...   
4 2023-12-30 20:58:12   Unknown  ...   

                                              source  \
0  {'uid': 'cdi00001985', 'schema': 'custom', 'na...   
1  {'uid': 'cdi00001985', 'schema': 'custom', 'na...   
2  {'uid': 'cdi00001985', 'schema': 'custom', 'na...   
3  {'uid': 'cdi00001985', 'schema': 'custom', 'na...   
4  {'uid': 'cdi00001985', 'schema': 'custom', 'na...   

                                             regions  \
0  [{'region_name': 'Americas', 'subregion_name':...   
1  [{'region_name': 'Americas', 'subregion_name':...   
2  [{'region_name': 'Americas', 'subregion_name':...   
3  [{'region_name': 'Americas', 'subregion_name':...   
4  [{'region_name': 'Americas', 'subregion_name':...   

                                             license  \
0  {'id': None, 'name': None, 'spdx': None, 'url'...   
1  {'id': None, 'name': None, 'spdx': None, 'url'...   
2  {'id': None, 'name': None, 'spdx': None, 'url'...   
3  {'id': None, 'name': None, 'spdx': None, 'url'...   
4  {'id': 'apache-2.0', 'name': 'Apache License 2...   

                                               usage scores  fields  \
0  {'size': '2491100', 'downloads': '42', 'likes'...   <NA>    <NA>   
1  {'size': None, 'downloads': '3', 'likes': '0',...   <NA>    <NA>   
2  {'size': '29604458', 'downloads': '31', 'likes...   <NA>    <NA>   
3  {'size': None, 'downloads': '43', 'likes': '0'...   <NA>    <NA>   
4  {'size': '712877', 'downloads': '16', 'likes':...   <NA>    <NA>   

                         year_range  \
0  {'y_start': None, 'y_end': None}   
1  {'y_start': None, 'y_end': None}   
2  {'y_start': None, 'y_end': None

### Проверка на полноту метаданных

Проверим, существуют ли датасеты без заголовков.

In [102]:
df_titlenan = conn.sql("""
SELECT * 
FROM hf_table
WHERE title IS NULL;
        """). df()

df_titlenan

Empty DataFrame
Columns: [id, cdi_id, int_id, title, short_text, description, url, date_created, date_changed, frequency, datatypes, formats, topics, geotopics, tags, has_archive, extracted_at, resource_cnt, resources, responsible, source, regions, license, usage, scores, fields, year_range, original, properties, num_resources]
Index: []

[0 rows x 30 columns]

Теперь посмотрим на заполненность описаний (description и short_text).

In [103]:
df_descnan = conn.sql("""
SELECT 
    COUNT(id)
FROM hf_table
WHERE description IS NULL;
        """). df()
df_descnan

count(id)
0     247057

In [104]:
df_shortnan = conn.sql("""
SELECT 
    COUNT(id)
FROM hf_table
WHERE short_text IS NULL;
        """). df()
df_shortnan

count(id)
0     247057

**247 057** датасетов не имеют никаких описаний. Это **68,02%**. Тут однозначно есть зависимость между заполненностью **description** и **short_text**. 



Проверим заполненность поля **topics**.
Не указано **topics** для **343 147** или **94.48%** не имеют указанных topics.
Напомню, что доля заполненности описаний выше.
То есть **topics** можно обогатить на основе этого поля и повысить заполненность до **32%**. 

In [105]:
df_topics = conn.sql("""
SELECT 
    topics,
    COUNT(id)
FROM hf_table
GROUP BY 1
ORDER BY 2 DESC;
        """). df()
df_topics

topics  count(id)
0                                                 <NA>     343147
1                        [Education culture and sport]       4637
2                                          [Transport]       2573
3                                        [Environment]       2272
4                             [Science and technology]       2047
..                                                 ...        ...
242  [Agriculture fisheries forestry and food,  Hea...          1
243  [Energy,  Justice legal system and public safe...          1
244  [Agriculture fisheries forestry and food,  Reg...          1
245  [Economy and finance,  Education culture and s...          1
246  [Population and society,  Science and technology]          1

[247 rows x 2 columns]

Заполненность **geotopics** хуже. Значения отсутствуют у **348 407** или **95.93%** датасетов.

In [106]:
df_geotopics = conn.sql("""
SELECT 
    geotopics,
    COUNT(id)
FROM hf_table
GROUP BY 1
ORDER BY 2 DESC;
        """). df()
df_geotopics

geotopics  count(id)
0                                                 <NA>     348407
1                                            [Society]       2300
2                            [Intelligence / Military]       2074
3                                            [Economy]       1687
4                          [Utilities / communication]       1124
..                                                 ...        ...
368       [Geoscientific information,  Transportation]          1
369                [Environment,  Society,  Structure]          1
370             [Boundaries,  Intelligence / Military]          1
371  [Structure,  Transportation,  Utilities / comm...          1
372  [Climatology / Meteorology / Atmosphere,  Geos...          1

[373 rows x 2 columns]

Посмотрим на заполненность поля **license**. 
**290 025** или **81,2%** не имеют указанной лицензии. 
Для остальных датасетов лидируют 
- Apache License 2.0 (**28 227 или 7.7%**)
- MIT License (**26 842 или 7.39%**)
- Creative Commons 4.0 (**5 231 или 1.44%**)

In [107]:
df_license = conn.sql("""
SELECT 
    (license::json->>'name') as license,
    COUNT(id)
FROM hf_table
GROUP BY 1
ORDER BY 2 DESC;
        """). df()
df_license

license  count(id)
0                                                None     290025
1                                  Apache License 2.0      28227
2                                         MIT License      26842
3      Creative Commons Attribution 4.0 International       5231
4   Creative Commons Attribution Share Alike 4.0 I...       2083
5   Creative Commons Attribution Non Commercial 4....       1835
6   Creative Commons Attribution Non Commercial Sh...       1769
7                Creative Commons Zero v1.0 Universal       1398
8   Creative Commons Attribution Non Commercial No...        875
9                          Academic Free License v3.0        787
10               GNU General Public License v3.0 only        784
11  Creative Commons Attribution Share Alike 3.0 U...        615
12        Do What The F*ck You Want To Public License        229
13  Creative Commons Attribution No Derivatives 4....        227
14                                      The Unlicense        223
15             GNU Affero General Public License v3.0        222
16          Creative Commons Attribution 3.0 Unported        199
17                               Artistic License 2.0        188
18           Creative Commons Attribution 2.0 Generic        184
19            BSD 3-Clause "New" or "Revised" License        156
20  Creative Commons Attribution Non Commercial 2....        153
21  Creative Commons Attribution Non Commercial Sh...        105
22    Community Data License Agreement Permissive 2.0         88
23               GNU General Public License v2.0 only         80
24  Creative Commons Attribution Non Commercial 3....         73
25       Community Data License Agreement Sharing 1.0         63
26  Creative Commons Attribution Non Commercial No...         60
27        GNU Lesser General Public License v3.0 only         58
28                  BSD 2-Clause "Simplified" License         53
29  Creative Commons Attribution Non Commercial Sh...         43
30                         Mozilla Public License 2.0         34
31                  European Union Public License 1.1         33
32                         Boost Software License 1.0         28
33                 Educational Community License v2.0         27
34           Creative Commons Attribution 2.5 Generic         26
35                            Etalab Open License 2.0         25
36                         BSD 3-Clause Clear License         23
37                           Microsoft Public License         21
38                                 PostgreSQL License         17
39    Community Data License Agreement Permissive 1.0         13
40                          Open Software License 3.0         12
41                                        ISC License          7
42                                       zlib License          7
43        GNU Lesser General Public License v2.1 only          7
44                          SIL Open Font License 1.1          6
45    University of Illinois/NCSA Open Source License          4
46                 LaTeX Project Public License v1.3c          4
47                         Eclipse Public License 1.0          3
48                         Eclipse Public License 2.0          3

In [108]:
df_license_topics = conn.sql(""" 
    SELECT
        (license::json->>'name') as license, 
        SUM(CASE WHEN topics IS NULL THEN 1 END) as "Число датасетов без topics",
        COUNT(id) as "Число датасетов",
        (SUM(CASE WHEN topics IS NULL THEN 1 END))/(COUNT(id))*100 as "Доля незаполненности"
    FROM hf_table
    GROUP BY 1
    ORDER BY 4 DESC;
""").df()

df_license_topics

license  \
0                          Eclipse Public License 1.0   
1                          Boost Software License 1.0   
2                  LaTeX Project Public License v1.3c   
3                                         ISC License   
4     University of Illinois/NCSA Open Source License   
5                          Eclipse Public License 2.0   
6                                  PostgreSQL License   
7     Community Data License Agreement Permissive 1.0   
8                                Artistic License 2.0   
9                                                None   
10                 Educational Community License v2.0   
11                                      The Unlicense   
12                         Academic Free License v3.0   
13                  BSD 2-Clause "Simplified" License   
14        Do What The F*ck You Want To Public License   
15  Creative Commons Attribution Non Commercial No...   
16        GNU Lesser General Public License v3.0 only   
17  Creative Commons Attribution Non Commercial Sh...   
18       Community Data License Agreement Sharing 1.0   
19                                        MIT License   
20  Creative Commons Attribution Share Alike 3.0 U...   
21  Creative Commons Attribution Non Commercial Sh...   
22                                 Apache License 2.0   
23               GNU General Public License v3.0 only   
24                                       zlib License   
25                           Microsoft Public License   
26        GNU Lesser General Public License v2.1 only   
27             GNU Affero General Public License v3.0   
28            BSD 3-Clause "New" or "Revised" License   
29                          Open Software License 3.0   
30                         Mozilla Public License 2.0   
31           Creative Commons Attribution 2.5 Generic   
32  Creative Commons Attribution Non Commercial 4....   
33     Creative Commons Attribution 4.0 International   
34           Creative Commons Attribution 2.0 Generic   
35               Creative Commons Zero v1.0 Universal   
36    Community Data License Agreement Permissive 2.0   
37                         BSD 3-Clause Clear License   
38  Creative Commons Attribution Non Commercial Sh...   
39          Creative Commons Attribution 3.0 Unported   
40  Creative Commons Attribution Non Commercial No...   
41  Creative Commons Attribution Non Commercial 3....   
42  Creative Commons Attribution Share Alike 4.0 I...   
43  Creative Commons Attribution No Derivatives 4....   
44               GNU General Public License v2.0 only   
45                  European Union Public License 1.1   
46                          SIL Open Font License 1.1   
47                            Etalab Open License 2.0   
48  Creative Commons Attribution Non Commercial 2....   

    Число датасетов без topics  Число датасетов  Доля незаполненности  
0                          3.0                3            100.000000  
1                         28.0               28            100.000000  
2                          4.0                4            100.000000  
3                          7.0                7            100.000000  
4                          4.0                4            100.000000  
5                          3.0                3            100.000000  
6                         17.0               17            100.000000  
7                         13.0               13            100.000000  
8                        184.0              188             97.872340  
9                     279660.0           290025             96.426170  
10                        26.0               27             96.296296  
11                       213.0              223             95.515695  
12                       732.0              787             93.011436  
13                        49.0               53             92.452830  
14                       211.0              229             92.139738  
15                        55.0             

Какие регионы и страны указываются при заполнении датасетов?
У **363 118** или **99.98%** датасетов указаны США.

In [109]:
df_regions = conn.sql("""
SELECT 
    (regions::json->0->>'region_name') as macroregions,
    (regions::json->0->>'country_or_area') as country,     
    COUNT(id)
FROM hf_table
GROUP BY 1, 2
ORDER BY 3 DESC;
        """). df()
df_regions

macroregions                   country  count(id)
0     Americas  United States of America     363118
1         None                      None         57

In [110]:
df_country_topics = conn.sql(""" 
    SELECT
        (regions::json->0->>'region_name') as macroregions,
        (regions::json->0->>'country_or_area') as country, 
        SUM(CASE WHEN topics IS NULL THEN 1 END) as "Число датасетов без topics",
        COUNT(id) as "Число датасетов",
        (SUM(CASE WHEN topics IS NULL THEN 1 END))/(COUNT(id))*100 as "Доля незаполненности"
    FROM hf_table
    GROUP BY 1, 2
    ORDER BY 5 DESC;
""").df()

df_country_topics

macroregions                   country  Число датасетов без topics  \
0     Americas  United States of America                    343097.0   
1         None                      None                        50.0   

   Число датасетов  Доля незаполненности  
0           363118             94.486365  
1               57             87.719298

Есть ли метаданные для заполнения временного периода, то есть указан ли временной отрезок за который собраны данные? Нет.
Также  эти метаданные не извлечь из каких-либо полей из источника.


In [111]:

df_year_range = conn.sql("""
SELECT 
    year_range,
    COUNT(id)
FROM hf_table
GROUP BY 1
ORDER BY 2 DESC;
        """). df()
df_year_range

year_range  count(id)
0  {'y_start': None, 'y_end': None}     363175

Проверим заполненность частоты обновлений датасетов (**frequency**). Она равна 0.

In [112]:
df_freq = conn.sql("""
SELECT 
    frequency,
    COUNT(id)
FROM hf_table
GROUP BY 1
ORDER BY 2 DESC;
        """). df()
df_freq

frequency  count(id)
0   Unknown     363175

Просмотр списка языков, которые указаны в метаданных датасетов.

In [113]:
df_original = conn.sql("""
SELECT 
    (original::json->>'languages') as languages,              
    COUNT(id)
FROM hf_table
GROUP BY 1
ORDER BY 2 DESC;
        """). df()
df_original


languages  count(id)
0                                                    []     324408
1                                                ["EN"]      20589
2                                                ["ZH"]       1203
3                                                ["RU"]        762
4                                                ["FR"]        737
...                                                 ...        ...
2507  ["ITA","FRA","FAS","DAN","ARZ","ARY","ARB","VI...          1
2508                              ["RU","EN","IT","DE"]          1
2509          ["ZH","EN","HI","ID","JA","RW","KO","ES"]          1
2510  ["EN","ZH","KO","FR","JA","ES","NO","ET","DE",...          1
2511  ["ASM","BEN","BRX","GUJ","HIN","KAN","KAS","KO...          1

[2512 rows x 2 columns]

Просмотр расширений файлов, указанных для датасетов.

In [114]:

df_extension = conn.sql("""
SELECT 
    (original::json->>'extensions') as ext,
    COUNT(id)
FROM hf_table
GROUP BY 1
ORDER BY 2 DESC;
        """). df()
df_extension

ext  count(id)
0      ["parquet"]     187121
1             [""]      88802
2         ["json"]      38513
3          ["csv"]      24096
4  ["imagefolder"]      10812
5  ["audiofolder"]       6656
6         ["text"]       4266
7   ["webdataset"]       1981
8        ["arrow"]        928

Проверим заполненность topics в разрезе расширений (форматов) файлов датасетов.

In [115]:
df_ext_topics = conn.sql(""" 
    SELECT
        (original::json->>'extensions') as "Расширение", 
        SUM(CASE WHEN topics IS NULL THEN 1 END) as "Число датасетов без topics",
        COUNT(id) as "Число датасетов",
        (SUM(CASE WHEN topics IS NULL THEN 1 END))/(COUNT(id))*100 as "Доля незаполненности"
    FROM hf_table
    GROUP BY 1
    ORDER BY 3 DESC;
""").df()

df_ext_topics

Расширение  Число датасетов без topics  Число датасетов  \
0      ["parquet"]                    176994.0           187121   
1             [""]                     84346.0            88802   
2         ["json"]                     36138.0            38513   
3          ["csv"]                     22175.0            24096   
4  ["imagefolder"]                     10238.0            10812   
5  ["audiofolder"]                      6498.0             6656   
6         ["text"]                      3989.0             4266   
7   ["webdataset"]                      1914.0             1981   
8        ["arrow"]                       855.0              928   

   Доля незаполненности  
0             94.587994  
1             94.982095  
2             93.833251  
3             92.027722  
4             94.691084  
5             97.626202  
6             93.506798  
7             96.617870  
8             92.133621

Проверим свойства датасетов (**properties**) в разрезе типов файлов и форматов данных.

In [116]:
df_properties = conn.sql("""
SELECT 
    (properties::json->>'modality') as properties,
    COUNT(id)
FROM hf_table
GROUP BY 1
ORDER BY 2 DESC;
        """). df()
df_properties

properties  count(id)
0                          [text]     170067
1                              []      86783
2                 [tabular, text]      40498
3                   [image, text]      19820
4                         [image]      16479
..                            ...        ...
89  [tabular, text, audio, video]          1
90    [image, audio, video, text]          1
91           [video, audio, text]          1
92            [timeseries, audio]          1
93     [audio, image, geospatial]          1

[94 rows x 2 columns]

In [117]:
df_prop_topics = conn.sql(""" 
    SELECT
        (properties::json->>'modality') as properties, 
        SUM(CASE WHEN topics IS NULL THEN 1 END) as "Число датасетов без topics",
        COUNT(id) as "Число датасетов",
        (SUM(CASE WHEN topics IS NULL THEN 1 END))/(COUNT(id))*100 as "Доля незаполненности"
    FROM hf_table
    GROUP BY 1
    ORDER BY 3 DESC;
""").df()

df_prop_topics

properties  Число датасетов без topics  \
0                           [text]                    159460.0   
1                               []                     83041.0   
2                  [tabular, text]                     38117.0   
3                    [image, text]                     18737.0   
4                          [image]                     15562.0   
..                             ...                         ...   
89            [video, audio, text]                         NaN   
90  [text, timeseries, geospatial]                         1.0   
91             [timeseries, audio]                         1.0   
92          [text, image, tabular]                         1.0   
93                  [video, image]                         1.0   

    Число датасетов  Доля незаполненности  
0            170067             93.763046  
1             86783             95.688096  
2             40498             94.120697  
3             19820             94.535822  
4             16479             94.435342  
..              ...                   ...  
89                1                   NaN  
90                1            100.000000  
91                1            100.000000  
92                1            100.000000  
93                1            100.000000  

[94 rows x 4 columns]

In [118]:
df_library = conn.sql("""
SELECT 
    (properties::json->>'library') as library,
    COUNT(id)
FROM hf_table
GROUP BY 1
ORDER BY 2 DESC;
        """). df()
df_library

library  count(id)
0             [datasets, pandas, mlcroissant, polars]     201340
1                                                  []      82133
2               [datasets, dask, mlcroissant, polars]      43484
3                             [datasets, mlcroissant]      29142
4                       [datasets, dask, mlcroissant]       2678
5                 [datasets, webdataset, mlcroissant]       1981
6   [datasets, pandas, mlcroissant, polars, distil...       1855
7    [datasets, pandas, mlcroissant, polars, argilla]        296
8                   [datasets, mlcroissant, fiftyone]         82
9   [datasets, dask, mlcroissant, polars, distilabel]         52
10                                         [fiftyone]         28
11                                       [distilabel]         22
12                                      [mlcroissant]         15
13                                       [webdataset]         10
14  [datasets, pandas, mlcroissant, polars, webdat...         10
15                                          [argilla]          8
16                    [datasets, pandas, mlcroissant]          8
17  [datasets, dask, mlcroissant, polars, webdataset]          7
18                                         [datasets]          4
19                [datasets, mlcroissant, webdataset]          3
20                                           [pandas]          3
21  [datasets, pandas, mlcroissant, polars, distil...          3
22     [datasets, dask, mlcroissant, polars, argilla]          2
23        [datasets, mlcroissant, webdataset, pandas]          1
24      [datasets, dask, mlcroissant, polars, pandas]          1
25                              [mlcroissant, pandas]          1
26                    [datasets, mlcroissant, pandas]          1
27                          [mlcroissant, webdataset]          1
28                    [datasets, mlcroissant, polars]          1
29            [datasets, dask, mlcroissant, fiftyone]          1
30                            [mlcroissant, datasets]          1
31                           [datasets, pandas, dask]          1

In [119]:
df_library_topics = conn.sql(""" 
    SELECT
        (properties::json->>'library') as library, 
        SUM(CASE WHEN topics IS NULL THEN 1 END) as "Число датасетов без topics",
        COUNT(id) as "Число датасетов",
        (SUM(CASE WHEN topics IS NULL THEN 1 END))/(COUNT(id))*100 as "Доля незаполненности"
    FROM hf_table
    GROUP BY 1
    ORDER BY 3 DESC;
""").df()

df_library_topics

library  \
0             [datasets, pandas, mlcroissant, polars]   
1                                                  []   
2               [datasets, dask, mlcroissant, polars]   
3                             [datasets, mlcroissant]   
4                       [datasets, dask, mlcroissant]   
5                 [datasets, webdataset, mlcroissant]   
6   [datasets, pandas, mlcroissant, polars, distil...   
7    [datasets, pandas, mlcroissant, polars, argilla]   
8                   [datasets, mlcroissant, fiftyone]   
9   [datasets, dask, mlcroissant, polars, distilabel]   
10                                         [fiftyone]   
11                                       [distilabel]   
12                                      [mlcroissant]   
13                                       [webdataset]   
14  [datasets, pandas, mlcroissant, polars, webdat...   
15                    [datasets, pandas, mlcroissant]   
16                                          [argilla]   
17  [datasets, dask, mlcroissant, polars, webdataset]   
18                                         [datasets]   
19                                           [pandas]   
20  [datasets, pandas, mlcroissant, polars, distil...   
21                [datasets, mlcroissant, webdataset]   
22     [datasets, dask, mlcroissant, polars, argilla]   
23                              [mlcroissant, pandas]   
24                            [mlcroissant, datasets]   
25                           [datasets, pandas, dask]   
26      [datasets, dask, mlcroissant, polars, pandas]   
27                    [datasets, mlcroissant, pandas]   
28        [datasets, mlcroissant, webdataset, pandas]   
29                          [mlcroissant, webdataset]   
30                    [datasets, mlcroissant, polars]   
31            [datasets, dask, mlcroissant, fiftyone]   

    Число датасетов без topics  Число датасетов  Доля незаполненности  
0                     189968.0           201340             94.351843  
1                      78287.0            82133             95.317351  
2                      40896.0            43484             94.048386  
3                      27478.0            29142             94.290028  
4                       2404.0             2678             89.768484  
5                       1914.0             1981             96.617870  
6                       1702.0             1855             91.752022  
7                        275.0              296             92.905405  
8                         76.0               82             92.682927  
9                         39.0               52             75.000000  
10                        26.0               28             92.857143  
11                        20.0               22             90.909091  
12                        12.0               15             80.000000  
13                         8.0               10             80.000000  
14                        10.0               10            100.000000  
15                         8.0                8            100.000000  
16                         3.0                8             37.500000  
17                         6.0                7             85.714286  
18                         1.0                4             25.000000  
19                         2.0                3             66.666667  
20                         2.0                3             66.666667  
21                         2.0                3             66.666667  
22                         2.0                2            100.000000  
23                         1.0                1            100.000000  
24                         1.0                1            100.000000  
25                         NaN                1                   NaN  
26                         1.0                1            100.000000  
27                         NaN                1                   NaN  
28                         1.0                1            100.000000  
29  

## Kaggle

Дата загрузки изучаемых файлов: 17 апреля 2025 (2025-04-17). 

In [77]:
#чтение и объединение всех parquet файлов в одну таблицу hf_table (из локальной папки)
conn.sql("""
CREATE TABLE kg_tbl AS
    SELECT * 
    FROM read_parquet('C:\\Users\\Kseniya\\dateno\\analyze may\\processed\\2025-04-17\\cdi00001986\\kaggle\\*.parquet', union_by_name=TRUE);
        """)

FloatProgress(value=0.0, layout=Layout(width='auto'), style=ProgressStyle(bar_color='black'))

In [ ]:
#просмотр таблицы в колоночном виде
conn.sql("""
    SELECT * 
    FROM kg_tbl;
        """)

┌──────────────────────────────────────────────────────────────────┬─────────────┬─────────┬──────────────────────────────────────────────────┬─────────────────────────────────────────────────────────────────────────────┬─────────────┬─────────────────────────────────────────────────────────────────────────────────────────────┬──────────────┬─────────────────────────┬───────────┬───────────┬───────────┬───────────┬───────────┬───────────┬─────────────┬──────────────┬──────────────┬─────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┬───────────────────────────────────────────────────┬────────────────────────

In [ ]:
#  подсчет числа строк в таблице
conn.sql("""
    SELECT
         COUNT(*) 
    FROM kg_tbl;
        """)

┌──────────────┐
│ count_star() │
│    int64     │
├──────────────┤
│       459874 │
└──────────────┘

Всего **459 874** датасетов содержится в таблице по Kaggle.

In [79]:
# преобразование таблицы в датафрейм и просмотр первых пяти строчек таблицы
df_kg = conn.sql("""
    SELECT *
    FROM kg_tbl;
        """).df()

df_kg.head()

id       cdi_id   int_id  \
0  9ce33da4721c6946ca6fa9c41ebd43e9205f710ece303b...  cdi00001986   126993   
1  35687dc8f2747aa1091b7d630d00bef46434714374b4ad...  cdi00001986   501269   
2  47bc783dba49f7cac9975ddb8972c23df51c662260128b...  cdi00001986  1431401   
3  da9ac08cb06ad1a6a30397dc2cea5ed7a85560943ff905...  cdi00001986  3878258   
4  f2a67974150c6d6c0f87a348aa34562f8393120c4f33e0...  cdi00001986  5217409   

                                  title  \
0                    ComputerProduction   
1                         KidneyChronic   
2                 Data Rumah 20 Teratas   
3  Statistical Learning | Data Analysis   
4                List of colleges in AP   

                                          short_text description  \
0                    An introduction to time series.               
1                                                                  
2                                                                  
3        Faculdade de Tecnologia da Baixada Santista               
4  A Comprehensive Overview of Higher Education I...               

                                                 url date_created  \
0  https://www.kaggle.com/datasets/nebulosito/com...          NaT   
1  https://www.kaggle.com/datasets/ur10ec065/kidn...          NaT   
2  https://www.kaggle.com/datasets/aliyahalmairah...          NaT   
3  https://www.kaggle.com/datasets/gabrielluizone...          NaT   
4  https://www.kaggle.com/datasets/upendra10/list...          NaT   

             date_changed frequency  ...  \
0 2019-02-25 19:41:33.283   Unknown  ...   
1 2020-02-05 00:30:04.227   Unknown  ...   
2 2021-06-25 15:36:37.993   Unknown  ...   
3 2023-10-18 21:48:29.560   Unknown  ...   
4 2024-06-15 15:53:12.017   Unknown  ...   

                                              source regions  \
0  {'uid': 'cdi00001986', 'schema': 'custom', 'na...    <NA>   
1  {'uid': 'cdi00001986', 'schema': 'custom', 'na...    <NA>   
2  {'uid': 'cdi00001986', 'schema': 'custom', 'na...    <NA>   
3  {'uid': 'cdi00001986', 'schema': 'custom', 'na...    <NA>   
4  {'uid': 'cdi00001986', 'schema': 'custom', 'na...    <NA>   

                                             license  \
0  {'id': None, 'name': None, 'spdx': None, 'url'...   
1  {'id': None, 'name': None, 'spdx': None, 'url'...   
2  {'id': None, 'name': None, 'spdx': None, 'url'...   
3  {'id': None, 'name': None, 'spdx': None, 'url'...   
4  {'id': None, 'name': None, 'spdx': None, 'url'...   

                                               usage scores  fields  \
0  {'size': '8837', 'downloads': '118', 'likes': ...   <NA>    <NA>   
1  {'size': '8548', 'downloads': '184', 'likes': ...   <NA>    <NA>   
2  {'size': '8425', 'downloads': '86', 'likes': '...   <NA>    <NA>   
3  {'size': '8816', 'downloads': '12', 'likes': '...   <NA>    <NA>   
4  {'size': '9741', 'downloads': '27', 'likes': '...   <NA>    <NA>   

                         year_range  \
0  {'y_start': None, 'y_end': None}   
1  {'y_start': None, 'y_end': None}   
2  {'y_start': None, 'y_end': None}   
3  {'y_start': None, 'y_end': None}   
4  {'y_start': None, 'y_end': None}   

                                            original properties num_resources  
0  {'topics': None, 'geotopics': None, 'regions':...       <NA>             1  
1  {'topics': None, 'geotopics': None, 'regions':...       <NA>             1  
2  {'topics': None, 'geotopics': None, 'regions':...       <NA>             1  
3  {'topics': None, 'geotopics': None, 'regions':...       <NA>             1  
4  {'topics': None, 'geotopics': None, 'regions':...       <NA>             1  

[5 rows x 30 columns]

### Проверка на полноту метаданных

Проверим, существуют ли датасеты без заголовков.

In [80]:
df_titlenan = conn.sql("""
SELECT * 
FROM kg_tbl
WHERE title IS NULL;
        """). df()

df_titlenan

Empty DataFrame
Columns: [id, cdi_id, int_id, title, short_text, description, url, date_created, date_changed, frequency, datatypes, formats, topics, geotopics, tags, has_archive, extracted_at, resource_cnt, resources, responsible, source, regions, license, usage, scores, fields, year_range, original, properties, num_resources]
Index: []

[0 rows x 30 columns]

Теперь посмотрим на заполненность описаний (description и short_text).

In [ ]:
df_descnan = conn.sql("""
SELECT 
    COUNT(id)
FROM kg_tbl
WHERE description IS NULL OR description = '';
        """). df()
df_descnan

count(id)
0     459874

Для всех **description** характерно отсутствие значений, но не типа Null, а в виде пустой строки ''. 

In [47]:
df_short_text = conn.sql("""
SELECT 
    short_text,
    COUNT(id)
FROM kg_tbl
GROUP BY 1
ORDER BY 2 DESC;
        """). df()
df_short_text

short_text  count(id)
0                                                            353710
1                              Stock Performance Dataset       4829
2                 Dr. Jagadish Tawade & Nitiraj Kulkarni       1895
3                  Explore Open Data from New York State        165
4           Explore Time Series from the Federal Reserve        157
...                                                  ...        ...
95398  Which cohort of leap day births is most repres...          1
95399  Areas from which Russia launched missiles and ...          1
95400          Scrapped laptops price from noon website           1
95401     List of One Day International Players of India          1
95402                     this is a Valorant gun dataset          1

[95403 rows x 2 columns]

При этом короткое описание (**short_text**) заполнено лучше. Значения отсутствуют для **353 710** или **76.9%** датасетов.


Поля **topics** и **geotopics** для данных Kaggle никак не заполнены. Поэтому ниже не будет проводится проверка доля заполненности topics в разных разрезах.

In [92]:
df_topics = conn.sql("""
SELECT 
    topics,
    COUNT(id)
FROM kg_tbl
GROUP BY 1
ORDER BY 2 DESC;
        """). df()
df_topics

topics  count(id)
0   <NA>     459874

In [93]:
df_geotopics = conn.sql("""
SELECT 
    geotopics,
    COUNT(id)
FROM kg_tbl
GROUP BY 1
ORDER BY 2 DESC;
        """). df()
df_geotopics

geotopics  count(id)
0      <NA>     459874

Проверим заполненность частоты обновлений датасетов (**frequency**). Она равна 0.

In [33]:
df_freq = conn.sql("""
SELECT 
    frequency,
    COUNT(id)
FROM kg_tbl
GROUP BY 1
ORDER BY 2 DESC;
        """). df()
df_freq

frequency  count(id)
0   Unknown     459874

Посмотрим на заполненность поля **license**. 
**361 842** или **78,68%** датасетов не имеют указанной лицензии. 
Для остальных датасетов лидируют 
- Apache License 2.0 (**58 928 или 12.81%**)
- MIT License (**30 094 или 6.54%**)
- Creative Commons Attribution Non Commercial Share (**4 884 или 1.06%**)

In [94]:
df_license = conn.sql("""
SELECT 
    (license::json->>'name') as license,
    COUNT(id)
FROM kg_tbl
GROUP BY 1
ORDER BY 2 DESC;
        """). df()
df_license

license  count(id)
0                                               None     361842
1                                 Apache License 2.0      58928
2                                        MIT License      30094
3  Creative Commons Attribution Non Commercial Sh...       4884
4  Creative Commons Attribution Share Alike 4.0 I...       3399
5  Creative Commons Attribution Share Alike 3.0 U...        727

Из поля **originals**, можно извлечь метаданные о расширении датасета/файла. Хотя и для **100%** датасетов - это zip архивы.

In [101]:

df_extensions = conn.sql("""
SELECT 
    (original::json->>'extensions') as ext,
    COUNT(id)
FROM kg_tbl
GROUP BY 1
ORDER BY 2 DESC;
        """). df()
df_extensions

ext  count(id)
0  [".zip"]     459874

Какие регионы и страны указываются при заполнении датасетов? Эти метаданные не заполнены у **100%** датасетов. 

In [35]:
df_regions = conn.sql("""
SELECT 
    (regions::json->0->>'region_name') as macroregions,
    (regions::json->0->>'country_or_area') as country,     
    COUNT(id)
FROM kg_tbl
GROUP BY 1, 2
ORDER BY 3 DESC;
        """). df()
df_regions

macroregions country  count(id)
0         None    None     459874

То же самое про временной период. Эти метаданные не заполнены у **100%** датасетов. 

In [38]:
df_year_range = conn.sql("""
SELECT 
    year_range,
    COUNT(id)
FROM kg_tbl
GROUP BY 1
ORDER BY 2 DESC;
        """). df()
df_year_range

year_range  count(id)
0  {'y_start': None, 'y_end': None}     459874

Не указаны также языки.

In [97]:
df_languages = conn.sql("""
SELECT 
    (original::json->>'languages') as lang,
    COUNT(id)
FROM kg_tbl
GROUP BY 1
ORDER BY 2 DESC;
        """). df()
df_languages

lang  count(id)
0  None     459874

Также нельзя извлечь дополнительные метаданные из поля **properties**.

In [100]:
df_properties = conn.sql("""
SELECT 
    properties,
    COUNT(id)
FROM kg_tbl
GROUP BY 1
ORDER BY 2 DESC;
        """). df()
df_properties

properties  count(id)
0       <NA>     459874